pyop2/fusion/extended.py

```diff
@@ -0,0 +1,706 @@
+                code += host.Kernel._ast_to_c(main, main_ast, main._opts)
+            else:
+                # AST not available so can't change the name, hopefully there
+                # will not be compile time clashes.
+                function_name = main._name
+                code += main._code
+            # Finally track the function name within this /fusion.Kernel/
+            for k in duplicates:
+                try:
+                    k._function_names[self.cache_key] = function_name
+                except AttributeError:
+                    k._function_names = {
+                        k.cache_key: k.name,
+                        self.cache_key: function_name
+                    }
+            code += "\n"
+        return code
+
+    def __init__(self, kernels, fused_ast=None, loop_chain_index=None):
+        """Initialize a :class:`fusion.Kernel` object.
+
+        :arg kernels: an iterator of some :class:`Kernel` objects. The objects
+            can be of class `fusion.Kernel` or of any superclass.
+        :arg fused_ast: the abstract syntax tree of the fused kernel. If not
+            provided, objects in ``kernels`` are considered "isolated C functions".
+        :arg loop_chain_index: index (i.e., position) of the kernel in a loop chain.
+            Meaningful only if ``fused_ast`` is specified.
+        """
+        # Protect against re-initialization when retrieved from cache
+        if self._initialized:
+            return
+        Kernel._globalcount += 1
+
+        # We need to distinguish between the kernel name and the function name(s).
+        # Since /fusion.Kernel/ are, in general, collections of functions, the same
+        # function (which is itself associated a Kernel) can appear in different
+        # /fusion.Kernel/ objects, but possibly under a different name (to avoid
+        # name clashes)
+        self._name = "_".join([k.name for k in kernels])
+        self._function_names = {self.cache_key: self._name}
+
+        self._cpp = any(k._cpp for k in kernels)
+        self._opts = dict(flatten([k._opts.items() for k in kernels]))
+        self._include_dirs = list(set(flatten([k._include_dirs for k in kernels])))
+        self._ldargs = list(set(flatten([k._ldargs for k in kernels])))
+        self._headers = list(set(flatten([k._headers for k in kernels])))
+        self._user_code = "\n".join(list(set([k._user_code for k in kernels])))
+        self._attached_info = {'fundecl': None, 'attached': False}
+
+        # What sort of Kernel do I have?
+        if fused_ast:
+            # A single AST (as a result of soft or hard fusion)
+            self._ast = fused_ast
+            self._code = self._ast_to_c(fused_ast)
+        else:
+            # Multiple functions (AST or strings, as a result of tiling)
+            self._ast = None
+            self._code = self._multiple_ast_to_c(kernels)
+        self._kernels = kernels
+
+        self._initialized = True
```

Comment:
I don't get anything from invoking the ``sequential.Kernel``'s constructor as I'd have to replace all fields with the union of the information from the fused kernels

Actually this led me to a good spot, I think. I was using inheritance to regenerate the cache key of the kernels being fused, but this is already stored in the cache_key attribute... I think I can get rid of the ``super`` in ``_cache_key``, thus making the instantiation of a ``fusion.Kernel`` much faster. 

Note that inheritance is still necessary as I'm using some of the superclass' methods